src/main/scala/elevate/core/strategies/predicate.scala

```diff
@@ -13,40 +15,31 @@ object predicate {
       case Success(_) => true
     }
 
-  case class not[P](s: Strategy[P]) extends Strategy[P] {
-    def apply(e: P): RewriteResult[P] = s(e) match {
-      case Success(_) => Failure(not(s))
-      case Failure(_) => Success(e)
-    }
-    override def toString: String = s"not($s)"
+  @combinator
+  def not[P]: Strategy[P] => Strategy[P] = s => e => s(e) match {
+    case Success(_) => Failure(not(s))
+    case Failure(_) => Success(e)
   }
 
   // only fails if the else case fails
   // id is required to not trigger the else case if the then case fails
-  def ifThenElse[P](p: Strategy[P],
-                    t: Strategy[P],
-                    e: Strategy[P]): Strategy[P] =
-    (p `;` (t <+ basic.id[P]())) <+ e
+  @combinator
+  def ifThenElse[P]: Strategy[P] => Strategy[P] => Strategy[P] => Strategy[P] =
+    p => t => e => (p `;` (t <+ basic.id)) <+ e
 
   def `if`[P]: Strategy[P] => Strategy[P] => Strategy[P] = p => t =>
-    ifThenElse(p,t,basic.id[P]())
+    ifThenElse(p)(t)(basic.id)
 
   def ![P](s: Strategy[P]): Strategy[P] = not(s)
 
-  case class isEqualTo[P](x: P) extends Strategy[P] {
-    def apply(p: P): RewriteResult[P] =
-      if (p == x) Success(p) else Failure(isEqualTo(x))
-    override def toString: String = s"isEqualTo($x)"
-  }
+  @strategy
+  def isEqualTo[P](x: P): Strategy[P] = p => if (p == x) Success(p) else Failure(isEqualTo(x))
 
   case class contains[P: Traversable](x: P) extends Strategy[P] {
```

Comment:
Shouldn't this be implemented using a macro as well?

Reply:
This shoiuld be added in the future.